src/utils/result.ts

```diff
@@ -0,0 +1,160 @@
+/**
+ * A Result wraps up a success state and a failure state, allowing you to
+ * return a single type from a function and discriminate between the two
+ * possible states in a principled way.
+ *
+ * Using it could look something like this:
+ *
+ * ```ts
+ * import { result } from '@utils';
+ *
+ * const mightFail = (input: number): Result<number, string> => {
+ *   try {
+ *     let value: number = calculateSomethingWithInput(input);
+ *     return result.ok(value);
+ *   } catch (e) {
+ *     return result.err(e.message);
+ *   }
+ * }
+ *
+ * const sumResult = mightFail(2);
+ *
+ * const msg = result.map(sumResult, (sum: number) => `the sum was: ${sum}`);
+ * ```
+ *
+ * A few utility methods are defined in this module, like `map` and `unwrap`,
+ * which are (probably obviously) inspired by the correspond methods on
+ * `std::result::Result` in Rust.
+ */
+export type Result<OnSuccess, OnFailure> = Ok<OnSuccess> | Err<OnFailure>;
+
+/**
+ * Type for the Ok state of a Result
+ */
+type Ok<T> = {
+  isOk: true;
+  isErr: false;
+  value: T;
+};
+
+/**
+ * Type for the Err state of a Result
+ */
+type Err<T> = {
+  isOk: false;
+  isErr: true;
+  value: T;
+};
```

Comment:
absolutely we could just have a single field to discriminate, I think TS just needs one boolean to discriminate between two variants like this - I included an `isErr` just in line with how this works in Rust (since this is basically off-brand `std::result::Result` haha)

we should also use a string like `variant: 'err' | 'ok'` if we wanted too

Reply:
I'd also argue that something like

```ts
if (foo.isErr) { }
```

is slightly more readable than the equivalent

```ts
if (!foo.isOk) { }
```

too easy to miss that little `!` haha

Reply:
Yeah, I'm not strictly opposed to keeping `isErr`, figured it'd be worth bringing up 😆